algorithms/src/snark/marlin/ahp/prover/oracles.rs

```diff
@@ -14,45 +14,72 @@
 // You should have received a copy of the GNU General Public License
 // along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
 
+use std::collections::BTreeMap;
+
 use snarkvm_fields::PrimeField;
 
-use crate::polycommit::sonic_pc::{LabeledPolynomial, LabeledPolynomialWithBasis};
+use crate::polycommit::sonic_pc::{LabeledPolynomial, LabeledPolynomialWithBasis, PolynomialInfo, PolynomialLabel};
 
 /// The first set of prover oracles.
 #[derive(Debug, Clone)]
 pub struct FirstOracles<'a, F: PrimeField> {
-    /// The evaluations of `Az`.
-    pub z_a: LabeledPolynomialWithBasis<'a, F>,
-    /// The evaluations of `Bz`.
-    pub z_b: LabeledPolynomialWithBasis<'a, F>,
+    pub(in crate::snark::marlin) batches: Vec<SingleEntry<'a, F>>,
     /// The sum-check hiding polynomial.
     pub mask_poly: Option<LabeledPolynomial<F>>,
+}
+
+impl<'a, F: PrimeField> FirstOracles<'a, F> {
+    /// Iterate over the polynomials output by the prover in the first round.
+    /// Intended for use when committing.
+    pub fn iter_for_commit(&'a self) -> impl Iterator<Item = LabeledPolynomialWithBasis<'a, F>> {
+        self.batches.iter().flat_map(|b| b.iter_for_commit()).chain(self.mask_poly.as_ref().map(Into::into))
+    }
+
+    /// Iterate over the polynomials output by the prover in the first round.
+    /// Intended for use when opening.
+    pub fn iter_for_open(&'a self) -> impl Iterator<Item = &'a LabeledPolynomial<F>> {
+        self.batches.iter().flat_map(|b| b.iter_for_open()).chain(self.mask_poly.as_ref())
+    }
+
+    pub fn matches_info(&self, info: &BTreeMap<PolynomialLabel, PolynomialInfo>) -> bool {
+        self.batches.iter().all(|b| b.matches_info(info))
+            && self.mask_poly.as_ref().map_or(true, |p| p.info() == &info[p.label()])
```

Comment:
```suggestion
            && self.mask_poly.as_ref().map_or(true, |p| Some(p.info()) == info.get(p.label()))
```
haven't tested it (might require a call to `.as_ref()` or an extra reference), but the suggestion won't panic if there is no `info[p.label()]`